testcases/cve/cve-2015-3290.c

```diff
@@ -0,0 +1,579 @@
+the start of espfix64's access to the original stack and when espfix64
+is done with the original stack.  If a nested NMI occurs during this
+window, then the atomic part of the basic nested NMI algorithm is
+observably non-atomic.
+
+Depending on exactly where in this window the nested NMI hits, the
+results vary.  Most nested NMIs will corrupt the return context and
+crash the calling process.  Some are harmless except that the nested NMI
+gets ignored.  There is a two-instruction window in which the return
+context ends up with user-controlled RIP and CS set to __KERNEL_CS.
+
+A careful exploit (attached) can recover from all the crashy failures
+and can regenerate a valid *privileged* state if a nested NMI occurs
+during the two-instruction window.  This exploit appears to work
+reasonably quickly across a fairly wide range of Linux versions.
+
+If you have SMEP, this exploit is likely to panic the system.  Writing
+a usable exploit against a SMEP system would be considerably more
+challenging, but it's surely possible.
+
+Measures like UDEREF are unlikely to help, because this bug is outside
+any region that can be protected using paging or segmentation tricks.
+However, recent grsecurity kernels seem to forcibly disable espfix64, so
+they're not vulnerable in the first place.
+
+A couple of notes:
+
+  - This exploit's payload just prints the text "CPL0".  The exploit
+    will keep going after printing CPL0 so you can enjoy seeing the
+    frequency with which it wins.  Interested parties could easily
+    write different payloads.  I doubt that any existing exploit
+    mitigation techniques would be useful against this type of
+    attack.
+
+  - If you are using a kernel older than v4.1, a 64-bit build of the
+    exploit will trigger a signal handling bug and crash.  Defenders
+    should not rejoice, because the exploit works fine when build
+    as a 32-bit binary or (so I'm told) as an x32 binary.
+
+  - This is the first exploit I've ever written that contains genuine
+    hexadecimal code.  The more assembly-minded among you can have
+    fun figuring out why :)
+
+[1] By "correctly", I mean that the register state ends up different
+from that which was saved in the stack frame, not that the
+implementation doesn't match the spec in the microcode author's minds.
+The spec is simply broken (differently on AMD and Intel hardware,
+perhaps unsurprisingly.)
+
+--Andy
+*/
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <inttypes.h>
+#include <asm/ldt.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <setjmp.h>
+#include <signal.h>
+#include <string.h>
+#include <sys/wait.h>
+#include "config.h"
+#if HAVE_PERF_EVENT_ATTR
+# include <linux/perf_event.h>
+#endif
+
+#include "tst_test.h"
+
+#if HAVE_PERF_EVENT_ATTR && (defined(__x86_64__) || defined(__i386__))
+
+#define tst_brkm(flags, cleanup, fmt, ...) do {		\
+		tst_brk(flags, fmt, ##__VA_ARGS__);	\
+	} while (0)
+#include "lapi/syscalls.h"
+
```

Comment:
You are supposed to use tst_syscall() in new code, the ltp_syscall() is deprecated as well as tst_brkm().